awesome/src/wayland_obj/xdg_shell.rs

```diff
@@ -102,103 +92,123 @@ impl XdgToplevel {
         FD: AsRawFd
     {
         let buffer = wayland_obj::create_buffer(fd.as_raw_fd(), size)?;
-        let state = unwrap_state_mut(&mut self.proxy);
+        let mut state = unwrap_state(self.as_ref());
         state.wl_surface.attach(Some(&buffer), 0, 0);
         state.size = size;
         state.buffer = Some(buffer);
+
         Ok(())
     }
 
     #[allow(dead_code)]
     pub fn commit(&self) {
-        unwrap_state(&self.proxy).wl_surface.commit();
+        unwrap_state(self.as_ref()).wl_surface.commit();
     }
 }
 
 impl Drop for XdgToplevel {
     fn drop(&mut self) {
         self.proxy.destroy();
-        unsafe {
-            let user_data = self.proxy.get_user_data() as *mut XdgToplevelState;
-            if !user_data.is_null() {
-                Box::from_raw(user_data);
-            }
-        }
+        // unsafe {
+        //     let user_data = self.proxy.get_user_data() as *mut XdgToplevelState;
+        //     if !user_data.is_null() {
+        //         Box::from_raw(user_data);
+        //     }
+        // }
+        // TODO(ried): leak? free user data?
```

Comment:
This can just be deleted I believe, since the user data is owned by the proxy object itself I think.

Reply:
I think so too - removed.